pkg/relayer/proxy/relay_meter.go

```diff
@@ -0,0 +1,384 @@
+// The relay reward is added optimistically, assuming that the relay will be volume / reward
+// applicable and the relay meter would remain up to date.
+func (rmtr *ProxyRelayMeter) AccumulateRelayReward(ctx context.Context, reqMeta servicetypes.RelayRequestMetadata) error {
+	rmtr.relayMeterMu.Lock()
+	defer rmtr.relayMeterMu.Unlock()
+
+	// Create a metric if it does not exist.
+	appMetrics, err := rmtr.ensureRequestAppMetrics(ctx, reqMeta, true)
+	if err != nil {
+		return err
+	}
+
+	// Get the cost of the relay based on the service and shared parameters.
+	relayCostCoin, err := getMinedRelayCostCoin(
+		appMetrics.sharedParams,
+		appMetrics.service,
+		appMetrics.serviceRelayDifficulty,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Increase the consumed stake amount by relay cost.
+	newConsumedCoin := appMetrics.consumedCoin.Add(relayCostCoin)
+
+	// If the consumed amount exceeds the max amount, return a rate limit error.
+	overServicingLimited := !rmtr.overServicingAllowanceCoins.IsNegative()
+	if overServicingLimited && appMetrics.maxCoin.IsLT(newConsumedCoin) {
+		return ErrRelayerProxyRateLimited.Wrapf(
+			"application has been rate limited, stake needed: %s, has: %s, ",
+			newConsumedCoin.String(),
+			appMetrics.maxCoin.String(),
+		)
+	}
+
+	appMetrics.consumedCoin = newConsumedCoin
+	return nil
+}
+
+// SetNonApplicableRelayReward updates the relay meter to make the relay reward for
+// the given relay request as non-applicable.
+// This is used when the relay is not volume / reward applicable but was optimistically
+// accounted for in the relay meter.
+func (rmtr *ProxyRelayMeter) SetNonApplicableRelayReward(ctx context.Context, reqMeta servicetypes.RelayRequestMetadata) error {
+	rmtr.relayMeterMu.Lock()
+	defer rmtr.relayMeterMu.Unlock()
+
+	appMetrics, err := rmtr.ensureRequestAppMetrics(ctx, reqMeta, false)
+	if err != nil {
+		return err
+	}
+
+	// Get the cost of the relay based on the service and shared parameters.
+	relayCost, err := getMinedRelayCostCoin(
+		appMetrics.sharedParams,
+		appMetrics.service,
+		appMetrics.serviceRelayDifficulty,
+	)
+	if err != nil {
+		return err
+	}
+
+	// Decrease the consumed stake amount by relay cost.
+	newConsumedAmount := appMetrics.consumedCoin.Sub(relayCost)
+
+	appMetrics.consumedCoin = newConsumedAmount
+	return nil
+}
+
+// forEachNewBlockFn is a callback function that is called every time a new block is committed.
+// It resets the relay meter's application stakes every new session so that new
+// application stakes can be metered.
+func (rmtr *ProxyRelayMeter) forEachNewBlockFn(ctx context.Context, block client.Block) {
+	rmtr.relayMeterMu.Lock()
+	defer rmtr.relayMeterMu.Unlock()
+
+	sharedParams, err := rmtr.sharedQuerier.GetParams(ctx)
+	if err != nil {
+		return
+	}
+	numBlocksPerSession := int64(sharedParams.GetNumBlocksPerSession())
+
+	// If the block observed is the last of the session, reset the relay meter's
```

Comment:
Is there a TODO here that needs to account for NumBlocksPerSession changing?

Reply:
Since `sharedParams` are fetched at each block, This should be fine. Although we must pay attention to cache invalidation when sharedQuerier will have its caching system.